text/0035-opentelemetry-protocol.md

```diff
@@ -0,0 +1,425 @@
+
+- `FailedNotRetryable` - processing of telemetry data failed. The client MUST NOT retry sending the same telemetry data. The telemetry data MUST be dropped. This for example can happen when the request contains bad data and cannot be deserialized or otherwise processed by the server. The client SHOULD maintain a counter of such dropped data.
+
+- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
+
+#### Throttling
+
+OTLP allows backpressure signalling.
+
+If the server is unable to keep up with the pace of data it receives from the client then it SHOULD signal that fact to the client. The client MUST then throttle itself to avoid overwhelming the server.
+
+To signal backpressure when using gRPC transport the server SHOULD return an error with code [Unavailable](https://godoc.org/google.golang.org/grpc/codes) and MAY supply additional [details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails) using [RetryInfo](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L40). Here is a sample Go code to illustrate:
+
+```go
+  // Do this on server side.
+  st, err := status.New(codes.Unavailable, "Server is unavailable").
+    WithDetails(&errdetails.RetryInfo{RetryDelay: &duration.Duration{Seconds: 30}})
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return st.Err()
+  
+  ...
+
+  // Do this on client side.
+  st := status.Convert(err)
+  for _, detail := range st.Details() {
+    switch t := detail.(type) {
+    case *errdetails.RetryInfo:
+      if t.RetryDelay.Seconds > 0 || t.RetryDelay.Nanos > 0 {
+        // Wait before retrying.
+      }
+    }
+  }
+```
+
+When the client receives this signal it SHOULD follow the recommendations outlined in documentation for `RetryInfo`:
+
+```
+// Describes when the clients can retry a failed request. Clients could ignore
+// the recommendation here or retry when this information is missing from error
+// responses.
+//
+// It's always recommended that clients should use exponential backoff when
+// retrying.
+//
+// Clients should wait until `retry_delay` amount of time has passed since
+// receiving the error response before retrying.  If retrying requests also
+// fail, clients should use an exponential backoff scheme to gradually increase
+// the delay between retries based on `retry_delay`, until either a maximum
+// number of retires have been reached or a maximum retry delay cap has been
+// reached.
+```
+
+The value of `retry_delay` is determined by the server and is implementation dependant. The server SHOULD choose a `retry_delay` value that is big enough to give the server time to recover, yet is not too big to cause the client to drop data while it is throttled.
+
```

Comment:
Do you want to cover the topic of which data should be dropped first? Older or newer.

Reply:
I think this should not be a protocol concern. The client may not even drop the data (e.g. it may have a large persistent queue that allows it to keep the data for a very long time). I believe the best behavior is highly dependent on the use case and should be a end-to-end system design concern that is orthogonal to the protocol (protocol only facilitates possible designs by allow the server to communicate this value to the client).